tests/csrf_tests/tests.py

```diff
@@ -652,45 +602,214 @@ def test_csrf_cookie_age_none(self):
+        """
+        A POST HTTPS request is accepted when USE_X_FORWARDED_PORT=True.
+        """
+        self._test_https_good_referer_behind_proxy()
 
-            POST = property(_get_post, _set_post)
+    @override_settings(ALLOWED_HOSTS=['www.example.com'], CSRF_COOKIE_DOMAIN='.example.com')
+    def test_https_good_referer_matches_cookie_domain(self):
+        """
+        A POST HTTPS request with a good referer should be accepted from a
+        subdomain that's allowed by CSRF_COOKIE_DOMAIN.
+        """
+        self._test_https_good_referer_matches_cookie_domain()
 
-        token = ('ABC' + self._csrf_id)[:CSRF_TOKEN_LENGTH]
+    @override_settings(ALLOWED_HOSTS=['www.example.com'], CSRF_COOKIE_DOMAIN='.example.com')
+    def test_https_good_referer_matches_cookie_domain_with_different_port(self):
+        """
+        A POST HTTPS request with a good referer should be accepted from a
+        subdomain that's allowed by CSRF_COOKIE_DOMAIN and a non-443 port.
+        """
+        self._test_https_good_referer_matches_cookie_domain_with_different_port()
 
-        req = CsrfPostRequest(token, raise_error=False)
-        resp = CsrfViewMiddleware().process_view(req, post_form_view, (), {})
-        self.assertIsNone(resp)
+    @override_settings(CSRF_COOKIE_DOMAIN='.example.com', DEBUG=True)
+    def test_https_reject_insecure_referer(self):
+        """
+        A POST HTTPS request from an insecure referer should be rejected.
+        """
+        req = self._get_POST_request_with_token()
+        req._is_secure_override = True
+        req.META['HTTP_REFERER'] = 'http://example.com/'
+        req.META['SERVER_PORT'] = '443'
+        response = CsrfViewMiddleware().process_view(req, post_form_view, (), {})
+        self.assertContains(
+            response,
+            'Referer checking failed - Referer is insecure while host is secure.',
+            status_code=403,
+        )
 
-        req = CsrfPostRequest(token, raise_error=True)
-        with patch_logger('django.security.csrf', 'warning') as logger_calls:
-            resp = CsrfViewMiddleware().process_view(req, post_form_view, (), {})
-            self.assertEqual(resp.status_code, 403)
-            self.assertEqual(logger_calls[0], 'Forbidden (%s): ' % REASON_BAD_TOKEN)
+
+class CsrfViewMiddlewareUseSessionsTest(CsrfViewMiddlewareTestMixin, SimpleTestCase):
+    """
+    This is a subclass of CsrfViewMiddlewareTest that changes its behaviour to work with
+    CSRF_USE_SESSIONS = True to test the session-based CSRF handling in the same way as the
+    cookie-based handling.
+    """
+
+    def _get_POST_bare_secret_csrf_cookie_request(self):
+        req = self._get_POST_no_csrf_cookie_request()
+        req.session[CSRF_SESSION_KEY] = self._csrf_id_cookie[:32]
+        return req
+
+    def _get_GET_csrf_cookie_request(self):
+        req = TestingHttpRequest()
+        req.session[CSRF_SESSION_KEY] = self._csrf_id_cookie
+        return req
+
+    def setUp(self):
```

Comment:
``@override_settings`` apparently works, I don't know why I didn't use it in the first place